tools/peerindex/peerindex.1.in
'\" et
.TH PEERINDEX 1 @VERSION@ PEERINDEX "User Commands"
.\"
.SH NAME
@UTILITY@
\(em inspect MRT TABLE_DUMPV2 Peer Index Table
.SH SYNOPSIS
.LP
.nf
@UTILITY@ \fB[\fIOPTIONS\fB]\fR... \fB[\fIFILES\fB]\fR...
.fi

.SH DESCRIPTION
The
.IR @UTILITY@
utility reads each possibly compressed Multithreaded Routing Toolkit
(MRT)
Routing Information Base
(RIB)
file specified by
.IR FILES ,
and extracts information related to PEER INDEX TABLE
records.
.IR @UTILITY@
processes and formats such records according to the options specified by
.IR OPTIONS .
.IR @UTILITY@
prints diagnostics to standard error,
it detects and tolerates data corruption as much as possible.
If data corruption cannot be tolerated, the entire MRT dump file is dropped,
.IR @UTILITY@
will then move to the next file in
.IR FILES ,
if any.
.P
Such events are always reported with reasonable diagnostic errors.
.P
See the
.IR EXAMPLES
section below for usage examples.

.SH OPTIONS
The following options are supported:

.IP "\fB\-o \fI<file>\fP" 10
Write output to
.BR file .
Instead of using standard output,
.IR @UTILITY@
shall output PEER INDEX TABLE information to the specified file. If option
occurs multiple times, last specified file is used.

.IP "\fB\-r or \-\-only\-refs\fP" 10
By default
.IR @UTILITY@
writes all PEER INDEX TABLE entries, this option causes
.IR @UTILITY@
to only print peers that are referenced by RIB entries.

.IP "\fB\-h or \-\-help\fP" 10
Prints a short help message, summarizing
.IR @UTILITY@
functionality.

.IP "\fB\-?\fP" 10
Equivalent to
.BR \-h .

.SH "LINE ORIENTED OUTPUT"
.IR @UTILITY@
splits each PEER INDEX TABLE record into multiple lines, every line represents a peer.
Note that
.IR OPTIONS
may cause some peers inside PEER INDEX TABLES to be discarded (See the
.IR OPTIONS
section for details).

Each peer is formatted as the following `|' separated fields:
.RS 4
.nf

PEER ENTRY|ASN32BIT
.fi
.P
.RE
.P
Fields have the following meaning:

.IP "\fBPEER ENTRY\fP" 4
The peer entry inside PEER INDEX TABLE, displayed as `peer\-address peer\-asn'.

.IP "\fBASN32BIT\fP" 4
May be either 1, if the peer AS number was encoded within 32 bits, or 0 otherwise.
.P
Note that this flag being 0 does not necessarily imply that the peer does not have
ASN32BIT capability, it is merely an indicator of how the peer entry was encoded within
the PEER INDEX TABLE record (namely, it only used 2 octets to encode its ASN).
.P
.RE

Multiple PEER INDEX TABLES are concatenated within the output, meaning that
the same peer appearing in multiple PEER INDEX TABLE records may appear twice in
.IR @UTILITY@
output.

.SH "EXIT STATUS"
The following exit values are returned:
.IP "\00" 6
All input data was scanned successfully,
and data was written to output correctly.
.IP >0 6
Errors were detected in input data, write error occurred,
or an unrecoverable error occurred (such as out of memory errors).

.SH STDIN
The standard input is used only if no
.IR FILES
arguments are provided, or when any of the specified
.IR FILES
arguments is `\-' , in which case MRT data is read from standard input at that
point, up to an <end\-of\-file>.
.P
Whenever
.IR @UTILITY@
reads from standard input, MRT data is assumed to be uncompressed.

.SH "INPUT FILES"
.IR @UTILITY@
supports most MRT dump formats as written by the majority of Route Collecting
projects (see the
.IR STANDARDS
section below for additional references).
MRT dumps may be provided either in their plain uncompressed form, or
as files compressed by
.IR gzip (1),
.IR bzip2 (1),
or
.IR xz (1).
.IR @UTILITY@
performs appropriate decompression on the fly.
File extension is used, in a case insensitive way, to discriminate among
supported compression formats. If the file extension is not recognized,
or there is no extension, then it is assumed to be uncompressed.

.SH STDOUT
Unless redirected explicitly via
.IR OPTIONS ,
the standard output is used to print a human readable text representation of
PEER INDEX TABLE contents, nothing else shall be written to the standard output.
.IR @UTILITY@
may detect and treat as error whenever the standard output is a regular file,
and is the same file as any of the
.IR FILES
arguments.
The default output format used by
.IR @UTILITY@
is documented in the
.IR "LINE ORIENTED OUTPUT"
section.

.SH STDERR
The standard error is used only for diagnostic messages and error reporting.

.SH EXAMPLES
This section contains some useful examples, demonstrating basic
.IR @UTILITY@
usage.

.IP \[bu]
The following is the simplest way to invoke
.IR @UTILITY@ :
.nf
\&
.in +2m
@UTILITY@
.in
\&
.fi
It formats and prints all peers inside the uncompressed MRT
RIB input data it receives from standard input.

.IP \[bu]
The following command:
.nf
\&
.in +2m
@UTILITY@ -r
.in
\&
.fi
formats and print only peer entries referenced by at least one RIB entry
inside the uncompressed MRT RIB input data coming from standard input.

.IP \[bu]
The following is equivalent to the previous example:
.nf
\&
.in +2m
@UTILITY@ -r ./rib.1.bz2
.in
\&
.fi
but takes MRT input data from a
.IR bzip (1)
compressed file.

.IP \[bu]
The following is equivalent to the previous example:
.nf
\&
.in +2m
@UTILITY@ -r ./rib.*.gz
.in
\&
.fi
but takes MRT input data from multiple
.IR gzip (1)
compressed files resulting from the glob expansion.
Most likely the command output is going to
contain data coming from multiple PEER INDEX TABLES,
peers may appear more than once.

.IP \[bu]
The following command:
.nf
\&
.in +2m
@UTILITY@ ./rib.*.gz
.in
\&
.fi
is similar to the previous example, but does not remove
unreferenced peer entries from output.

.SH SEE ALSO
.IR awk (1),
.IR grep (1)

.SH STANDARDS
The
.IR @UTILITY@
utility conforms to:
.IP \[bu] 2m
.IR "RFC 6396" " \-" "Multi\-Threaded Routing Toolkit (MRT) Routing Information Export Format"
.IP \[bu] 2m
.IR "RFC 8050" " \- " "Multi\-Threaded Routing Toolkit (MRT) Routing Information Export Format with BGP Additional Path Extensions"

.SH AUTHOR
.IR @UTILITY@
was written by
.UR lcg@\:inventati.\:org
Lorenzo Cogotti
.UE .


doc/HISTORY.md
Micro BGP Suite history
=======================

This document summarizes a bit of history of the project, it tells you anything
you didn't want to know about `ubgpsuite` and never cared to ask.

`ubgpsuite` was created as an evolution over `bgpscanner` and its companion
library, `isocore`. `bgpscanner` was originally developed by me starting in 2017
as part of the Isolario Project, under the Institute of Informatics and
Telematics of the Italian National Research Council
[IIT-CNR](https://www.iit.cnr.it/).
`bgpscanner` was covered by the MIT license terms and is still available
(as of May 2021) at Isolario's
[website](https://isolario.it/web_content/php/site_content/tools.php).
Despite the fact that `bgpscanner` code was developed mostly by me, the
help of the Isolario team, their patience, and their knowledge about BGP's
nuts and bolts was invaluable to get it rolling.

By mid 2019, my collaboration with IIT-CNR has ceased.
In an attempt to further the concepts behind `bgpscanner` and keep experimenting
with them, I started the `ubgpsuite` project. At this time I was involved with
[Alpha Cogs](https://www.alphacogs.com), a company I co-founded, so I undertook
`ubgpsuite` development and got it moving forward with its financial support.

`ubgpsuite` was born by a partial rewrite of `bgpscanner`, taking advantage of
the fact that there was no more interoperability constrain with a larger
project -- `isocore` was originally intended to be used by other
components inside the Isolario project as well, but that wasn't the
case anymore. This allowed some minor improvement to the codebase, but the
overall software architecture was unchanged. Though `ubgpsuite` was relicensed
under the terms of LGPLv3+ for library code and GPLv3+ for utilities and tools.
The choice was motivated by my intention to keep the project free
(as in freedom) forever, considering that the only reason I was able to continue
developing the code I authored at IIT-CNR and keep the project alive was
its original open source license.

Unfortunately a chronical lack of time, due to more pressing company priorities,
has hit the fan during that year and most of 2020. But finally,
by the end of 2020, I got the chance to dedicate some time to the project.
As a matter of fact, I left Alpha Cogs and devoted a bit of myself to
move `ubgpsuite` and other projects I cared about out of stagnation.
`ubgpsuite` was then outside Alpha Cogs and became the first of
DoubleFourteen Code Forge projects -- 1414° for short, an initiative I
founded to research and develop free software, in the hope of improving
the software ecosystem.

On 2021, a total rewrite of `ubgpsuite` was complete and ready for release,
with this document included. History of the project will thereafter remain
unread.

Keep developing the code you love and enjoy,

---

Lorenzo Cogotti


tools/peerindex/peerindex.c
// SPDX-License-Identifier: GPL-3.0-or-later

/**
 * \file peerindex.c
 *
 * `peerindex` main entry point and general command line parsing.
 *
 * \copyright The DoubleFourteen Code Forge (C) All Rights Reserved
 * \author    Lorenzo Cogotti
 */

#include "peerindex_local.h"

#include "bgp/bytebuf.h"
#include "cpr/flate.h"
#include "cpr/bzip2.h"
#include "cpr/xz.h"
#include "sys/endian.h"
#include "sys/fs.h"
#include "sys/con.h"
#include "sys/sys.h"
#include "argv.h"
#include "bufio.h"
#include "strlib.h"

#include <assert.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>

#define BYTEBUFSIZ (256 * 1024)

typedef enum {
	ONLY_REFS_FLAG,
	OUTPUT_FLAG,

	NUM_FLAGS
} PeerindexOpt;

static Optflag options[] = {
	[ONLY_REFS_FLAG] = {
		'r', "only-refs", NULL, "Only dump peers referenced by RIBs", ARG_NONE
	},
	[OUTPUT_FLAG] = {
		'o', NULL, "file", "Write output to file", ARG_REQ
	},

	[NUM_FLAGS] = { '\0' }
};

static PeerindexState S;
static BGP_FIXBYTEBUF(BYTEBUFSIZ) bgp_msgBuf = { BYTEBUFSIZ };

static void Peerindex_SetupCommandLine(char *argv0)
{
	Sys_StripPath(argv0, NULL);
	Sys_StripFileExtension(argv0, NULL);

	com_progName   = argv0;
	com_synopsis   = "[OPTIONS...] [FILES...]";
	com_shortDescr = "MRT TABLE_DUMPV2 Peer Index Table inspection tool";
	com_longDescr  =
		"Reads MRT dumps in various formats, and inspects the\n"
		"MRT TABLE_DUMPV2 peer index table, and prints it to stdout.\n"
		"If - is found inside FILES, then input is read from stdin.\n"
		"If no FILES arguments are provided, then input\n"
		"is implicitly expected from stdin.\n"
		"Any diagnostic message is logged to stderr.";
}

static void Peerindex_Fatal(const char *fmt, ...)
{
	va_list va;

	Sys_Print(STDERR, com_progName);
	Sys_Print(STDERR, ": ERROR: ");

	va_start(va, fmt);
	Sys_VPrintf(STDERR, fmt, va);
	va_end(va);

	Sys_Print(STDERR, "\n");
	exit(EXIT_FAILURE);
}

static void Peerindex_Warning(const char *fmt, ...)
{
	va_list va;

	Sys_Print(STDERR, com_progName);
	Sys_Print(STDERR, ": ");
	if (S.filename) {
		Sys_Print(STDERR, S.filename);
		Sys_Print(STDERR, ": ");
	}

	Sys_Print(STDERR, "WARNING: ");

	va_start(va, fmt);
	Sys_VPrintf(STDERR, fmt, va);
	va_end(va);

	Sys_Print(STDERR, "\n");
}

void Peerindex_DropRecord(const char *fmt, ...)
{
	va_list va;

	Sys_Print(STDERR, com_progName);
	Sys_Print(STDERR, ": ");

	assert(S.filename);
	Sys_Print(STDERR, S.filename);
	Sys_Print(STDERR, ": ");

	va_start(va, fmt);
	Sys_VPrintf(STDERR, fmt, va);
	va_end(va);

	Sys_Print(STDERR, "\n");
	S.nerrors++;

	// If we're dropping a record we have to kill S.rec
	Bgp_ClearMrt(&S.rec);

	// ...but don't drop PEER_INDEX_TABLE

	longjmp(S.dropRecordFrame, 1);
}

void Peerindex_DropFile(const char *fmt, ...)
{
	va_list va;

	Sys_Print(STDERR, com_progName);
	Sys_Print(STDERR, ": ");

	assert(S.filename);
	Sys_Print(STDERR, S.filename);
	Sys_Print(STDERR, ": ");

	va_start(va, fmt);
	Sys_VPrintf(STDERR, fmt, va);
	va_end(va);

	Sys_Print(STDERR, "\n");
	S.nerrors++;

	// Must clear PEER_INDEX_TABLE also, along with S.rec...
	Bgp_ClearMrt(&S.rec);
	Bgp_ClearMrt(&S.peerIndex);
	S.hasPeerIndex = FALSE;

	// Drop any opened file and jump out
	if (S.infOps) {
		if (S.infOps->Close) S.infOps->Close(S.inf);

		S.inf    = NULL;
		S.infOps = NULL;
	}
	S.filename = NULL;

	longjmp(S.dropFileFrame, 1);
}

NOINLINE static void Peerindex_HandleError(BgpRet err, Srcloc *loc, void *obj)
{
	USED(obj);

	// On out of memory we die with an appropriate backtrace
	if (err == BGPENOMEM) {
		loc->call_depth++;  // include this function itself

		_Sys_OutOfMemory(loc->filename, loc->func, loc->line, loc->call_depth);
	}
	// On I/O error we drop the entire file
	if (err == BGPEIO)
		Peerindex_DropFile("I/O error while reading MRT dump");  // TODO: better diagnostics

	// Hopefully we are not dealing with a programming error...
	assert(err != BGPEBADMRTTYPE);
	assert(BGP_ISMRTERR(err));

	if (err == BGPEBADPEERIDXCNT || err == BGPEBADRIBV2CNT) {
		// Only warrant a warning
		Peerindex_Warning("CORRUPT MRT RECORD: %s", Bgp_ErrorString(err));
		S.nerrors++;
		return;
	}

	// Anything else requires a full record drop
	Peerindex_DropRecord("SKIPPING MRT RECORD: %s", Bgp_ErrorString(err));
}

static void Peerindex_ApplyProgramOptions(void)
{
	if (options[ONLY_REFS_FLAG].flagged)
		S.peerIndexClearVal = 0;
	else
		S.peerIndexClearVal = 0xff;  // so we always print the full table

	if (options[OUTPUT_FLAG].flagged) {
		const char *filename = options[OUTPUT_FLAG].optarg;

		Fildes fd = Sys_Fopen(filename, FM_WRITE, /*hints=*/0);
		if (fd == FILDES_BAD)
			Peerindex_Fatal("Can't open output file \"%s\"", filename);

		S.outf    = STM_FILDES(fd);
		S.outfOps = Stm_FildesOps;
	} else {
		S.outf    = STM_CONHN(STDOUT);
		S.outfOps = Stm_ConOps;
	}
}

static void Peerindex_Init(void)
{
	S.rec.allocp = &bgp_msgBuf;
	S.rec.memOps = Mem_BgpBufOps;
}

static void Peerindex_OpenMrtDump(const char *filename)
{
	S.filename = filename;
	if (strcmp(S.filename, "-") == 0) {
		// Direct read from stdin, assume uncompressed
		S.inf    = STM_FILDES(CON_FILDES(STDIN));
		S.infOps = Stm_NcFildesOps;
		return;
	}

	Fildes fh = Sys_Fopen(filename, FM_READ, FH_SEQ);
	if (fh == FILDES_BAD)
		Peerindex_DropFile("Can't open file");

	const char *ext = Sys_GetFileExtension(filename);
	if (Df_stricmp(ext, ".bz2") == 0) {
		S.inf    = Bzip2_OpenDecompress(STM_FILDES(fh), Stm_FildesOps, /*opts=*/NULL);
		S.infOps = Bzip2_StmOps;

		Bzip2Ret err = Bzip2_GetErrStat();
		if (err)
			Peerindex_DropFile("Can't read Bz2 archive: %s", Bzip2_ErrorString(err));

	} else if (Df_stricmp(ext, ".gz") == 0 || Df_stricmp(ext, ".z") == 0) {
		S.inf    = Zlib_InflateOpen(STM_FILDES(fh), Stm_FildesOps, /*opts=*/NULL);
		S.infOps = Zlib_StmOps;

		ZlibRet err = Zlib_GetErrStat();
		if (err)
			Peerindex_DropFile("Can't read Zlib archive: %s", Zlib_ErrorString(err));

	} else if (Df_stricmp(ext, ".xz") == 0) {
		S.inf    = Xz_OpenDecompress(STM_FILDES(fh), Stm_FildesOps, /*opts=*/NULL);
		S.infOps = Xz_StmOps;

		XzRet err = Xz_GetErrStat();
		if (err)
			Peerindex_DropFile("Can't read LZMA archive: %s", Xz_ErrorString(err));

	} else {
		// Assume uncompressed file
		S.inf    = STM_FILDES(fh);
		S.infOps = Stm_FildesOps;
	}
}

static void Peerindex_MarkPeerRefs(void)
{
	Mrtribiterv2 it;
	Mrtribentv2 *rib;

	Bgp_StartMrtRibEntriesv2(&it, &S.rec);
	while ((rib = Bgp_NextRibEntryv2(&it)) != NULL) {
		Uint16 idx = beswap16(rib->peerIndex);
		if (idx >= S.npeers) {
			Peerindex_Warning("CORRUPT MRT RECORD: Peer index '%u' is out of range", (unsigned) idx);
			S.nerrors++;
			continue;
		}

		MARKPEERINDEX(S.peerIndexRefs, idx);
	}
}

static void Peerindex_FlushPeerIndexTable(void)
{
	char buf[IPV6_STRLEN + 1];
	Stmbuf sb;

	Ipadr adr;

	Mrtpeeriterv2 it;
	Mrtpeerentv2 *peer;

	if (!S.hasPeerIndex)
		return;  // NOP

	Uint16 idx = 0;

	Bufio_Init(&sb, S.outf, S.outfOps);

	Bgp_StartMrtPeersv2(&it, &S.peerIndex);
	while ((peer = Bgp_NextMrtPeerv2(&it)) != NULL) {
		if (ISPEERINDEXREF(S.peerIndexRefs, idx)) {
			Asn asn    = MRT_GETPEERADDR(&adr, peer);
			char *eptr = Ip_AdrToString(&adr, buf);

			Bufio_Putsn(&sb, buf, eptr - buf);
			Bufio_Putc(&sb, ' ');
			Bufio_Putu(&sb, beswap32(ASN(asn)));
			Bufio_Putc(&sb, '|');
			Bufio_Putc(&sb, ISASN32BIT(asn) ? '1' : '0');
			Bufio_Putc(&sb, '\n');
		}

		idx++;
	}
	Bufio_Flush(&sb);
}

static void Peerindex_ProcessRecord(void)
{
	const Mrthdr *hdr = MRT_HDR(&S.rec);
	if (hdr->type != MRT_TABLE_DUMPV2)
		return;  // don't care for anything else

	if (hdr->subtype == TABLE_DUMPV2_PEER_INDEX_TABLE) {
		// Dump peers seen so far, if any
		Peerindex_FlushPeerIndexTable();
		// Clear previous PEER_INDEX_TABLE, if any
		Bgp_ClearMrt(&S.peerIndex);
		// Clear reference table...
		memset(S.peerIndexRefs, S.peerIndexClearVal, sizeof(S.peerIndexRefs));

		//...and update current PEER_INDEX_TABLE
		size_t npeers;

		MRT_MOVEREC(&S.peerIndex, &S.rec);
		Bgp_GetMrtPeerIndexPeers(&S.peerIndex, &npeers, /*nbytes=*/NULL);
		S.npeers       = npeers;
		S.hasPeerIndex = TRUE;
		return;
	}

	if (!TABLE_DUMPV2_ISRIB(hdr->subtype))
		return;  // irrelevant to us

	if (!S.hasPeerIndex)
		Peerindex_DropRecord("SKIPPING TABLE_DUMPV2 RECORD - No PEER_INDEX_TABLE found yet");

	Peerindex_MarkPeerRefs();
}

static void Peerindex_ProcessMrtDump(const char *filename)
{
	// NOTE: This call is responsible to set:
	// S.filename, S.inf, S.infOps.
	// These must be cleared either here or within a file drop.
	Peerindex_OpenMrtDump(filename);

	setjmp(S.dropRecordFrame);  // NOTE: The ONLY place where this is set
	while (Bgp_ReadMrt(&S.rec, S.inf, S.infOps) == OK) {
		Peerindex_ProcessRecord();

		Bgp_ClearMrt(&S.rec);
	}

	Peerindex_FlushPeerIndexTable();  // execute the very last dump

	// Don't need PEER_INDEX_TABLE anymore
	Bgp_ClearMrt(&S.peerIndex);
	S.hasPeerIndex = FALSE;

	// Finally close file and call it a day
	if (S.infOps->Close) S.infOps->Close(S.inf);

	S.inf      = NULL;
	S.infOps   = NULL;
	S.filename = NULL;
}

int main(int argc, char **argv)
{
#ifndef NDEBUG
	Sys_SetErrFunc(SYS_ERR_ABORT, NULL);
#else
	Sys_SetErrFunc(SYS_ERR_QUIT, NULL);
#endif
	Bgp_SetErrFunc(Peerindex_HandleError, NULL);

	Peerindex_SetupCommandLine(argv[0]);

	int optind = Com_ArgParse(argc, argv, options, /*flags=*/0);
	if (optind == OPT_HELP)
		return EXIT_SUCCESS;  // only called to get help...
	if (optind < 0)
		return EXIT_FAILURE;  // can't parse command line

	Peerindex_ApplyProgramOptions();

	Peerindex_Init();  // initialize according to command line

	// Done with options
	argc -= optind;
	argv += optind;

	// Move to filtering rules and compile them
	if (argc == 0) {
		// If no FILES are provided, read from stdin
		static const char *const stdinFile[] = { "-" };

		argv = (char **) stdinFile;
		argc = ARRAY_SIZE(stdinFile);
	}

	volatile int i = 0;

	setjmp(S.dropFileFrame);  // NOTE: The ONLY place where this is set
	while (i < argc)
		Peerindex_ProcessMrtDump(argv[i++]);

	if (S.outfOps->Close) S.outfOps->Close(S.outf);

	return (S.nerrors > 0) ? EXIT_FAILURE : EXIT_SUCCESS;
}
